File: Concurrency/P1114_Print_in_Order/Foo.go
package main

import (
	"fmt"
	// "time"
)

type Foo struct {
	ch1 chan struct{}
	ch2 chan struct{}
	ch3 chan struct{}
}

func newFoo() Foo {
	f := Foo{
		ch1: make(chan struct{}),
		ch2: make(chan struct{}),
		ch3: make(chan struct{}),
	}
	return f
} 

func (this *Foo) first() {
	fmt.Println("first")
	this.ch1 <- struct{}{}
}

func (this *Foo) second() {
	<- this.ch1
	fmt.Println("second");
	this.ch2 <- struct{}{}
}

func (this *Foo) third() {
	<- this.ch2
	fmt.Println("third");
	this.ch3 <- struct{}{}
}

func main() {
	f := newFoo()
	go func() {
		go f.first()
		go f.second()
		go f.third()
	}()
	<- f.ch3
}

File: Concurrency/P1114_Print_in_Order/Foo.java
/**
 * 1114. Print in Order -- Easy
 */

class Foo {

    private Object lock = new Object();
    private boolean firstDone = false;
    private boolean secondDone = false;

    public Foo() {
        
    }

    public void first(Runnable printFirst) throws InterruptedException {
        synchronized(lock) {
            // printFirst.run() outputs "first". Do not change or remove this line.
            printFirst.run();
            firstDone = true;
            lock.notifyAll();
        }
    }

    public void second(Runnable printSecond) throws InterruptedException {
        synchronized(lock) {
            while (!firstDone) {
                lock.wait();
            }
            // printSecond.run() outputs "second". Do not change or remove this line.
            printSecond.run();
            secondDone = true;
            lock.notifyAll();
        }
    }

    public void third(Runnable printThird) throws InterruptedException {
        synchronized(lock) {
            while (!secondDone) {
                lock.wait();
            }

            // printThird.run() outputs "third". Do not change or remove this line.
            printThird.run();
        }
        
    }

    // public static void main(String[] args) {
    //     Foo foo = new Foo();

    // }
}